docs/design/provisioning.md

```diff
@@ -0,0 +1,328 @@
+
+`PersistentVolumeRecycler` is a plugin that reclaims discarded resources.
+
+### Goals
+
+* Allow users to request volumes via selector [#14908](https://github.com/kubernetes/kubernetes/issues/14908)
+* Allow admins to configure many provisioners of persistent volumes
+* Allow admins to configure many recyclers of persistent volumes
+* Decouple provisioning code from volume plugin code
+
+
+### PersistentStorageConfig
+
+Administrators map binders, provisioners, and recyclers to specific types of persistent claims and volumes by creating a PersistentStorageConfig file.  The config file is set as a CLI flag.
+
+Binders contains two selectors against which they match claims for binding.  One selector matches claims and the other selector matches volumes.  Both must be true in order to bind, in addition to other matching semantics (e.g, capacity). Claims can remain unbound indefinitely.
+
+Provisioners create resources on demand for claims matching its selector. PersistentVolumeClaims with matching labels will be dynamically provisioned by a plugin named in the config.  Volume attributes are copied from a template.  A specific security context can be assigned to the provisioner.  Volumes created by the provisioner automatically bind to the claim for which it was provisioned.
+
+Recyclers have a selector against which they match persistent volumes for recycling.  A specific security context can be assigned to the recycler. Volumes are in Released state after their bound claims are deleted.  Recyclers may delete volumes or reset phase back to Available, according to each plugin.  Volumes without a matching recycler may be stay Released indefinitely.
+
+
+```
+
+// PersistentStorageConfig is the config defining the management of persistent volumes.
+type PersistentStorageConfig struct {
+	Recyclers    []RecyclerConfig
+	Provisioners []ProvisionerConfig
+}
+
+// RecyclerConfig is a label set for persistent volumes, a plugin w/ optional arguments, and a security context to work
+// within. The Recycler controller running in master will use the specified plugin and optional arguments.
+// Recyclers can be plugins that delete cloud volumes through APIs or pods that scrub a volume for reuse.
+// Each Recycler plugin requires its own documentation for usage and plugin arguments.
+type RecyclerConfig struct {
+	// PersistentVolumeSelector is a selector which must be true for the recycler to match the persistent volume
+	PersistentVolumeSelector Selector
+	// RecyclerPluginName is the name of the plugin that can recycle the matched persistent volume.
+	RecyclerPluginName string
+	// Args is a map of key/value pairs passed to the driver (but not plugin) as arguments
+	Args *map[string]string
+	// SecurityContextName is the name of the security context to use when performing recycling operations
+	SecurityContextName string
```

Comment:
@markturansky 

I don't understand what this represents, since security contexts are not resources and cannot be looked up by name.